Fixes #28900 - Add search to Taxonomies in top navigation dropdowns

stejskalleos opened the pull request:
**Description of problem:**
When a Satellite has many organizations and locations it can be time consuming to search and select the one you want from the context menu. I would like to be able to enter a name in a search field, and for auto-complete to be supported so that I can search by typing in the first part of the name.

**Fix**
If there are more than six Organizations or Locations in the top main menu, search input is displayed. Users then can filter items in dropdown (by it's name).

If there is six or less Org/Loc, search field is hidden.

stejskalleos committed:
Fixes #28900 - Add search to Taxonomies in top navigation dropdowns

xprazak2 commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js:
```diff
@@ -3,63 +3,113 @@ import PropTypes from 'prop-types';
+class TaxonomyDropdown extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { filtered: this.props.taxonomies };
```
Setting prop values to state is generally not recommended:
https://reactjs.org/docs/react-component.html#constructor
https://vasanthk.gitbooks.io/react-bits/anti-patterns/01.props-in-initial-state.html

I'd go the other way around:
* have `searchValue` in component state
* split `handleChange` into 2 functions 
  * hander that changes `searchValue` in state on input change
  * function that filters `taxonomies` based on `searchValue`

At this point, you should be able to remove `componentWillReceiveProps` as you no longer need to watch props for changes to keep state in sync.

You can also consider using [react hooks](https://reactjs.org/docs/hooks-intro.html) instead of a class.

stejskalleos replied:
Updated.

xprazak2 commented:
Nice feature, it needs a bit more polish.

The `form-inline` class makes the dropdown wider than it needs to be:

![form-inline](https://user-images.githubusercontent.com/2664090/74329027-4e725800-4d8f-11ea-833c-0ae03467370f.png)

and it contracts as I start to type:

![form-inline-cont](https://user-images.githubusercontent.com/2664090/74329105-7b266f80-4d8f-11ea-8e4d-527d90eb5c6e.png)

xprazak2 commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js:
```diff
@@ -4,62 +4,98 @@ import NavItem from './NavItem';
+      const { searchValue } = this.state;
+
+      if (searchValue === '') {
+        return this.props.taxonomies;
```
nit: you already use destructuring above to get taxonomies from props so you can take advantage of that and pass them as an arg to your filter function - that way you do not need to pull taxonomies out of the props here:

```js
const filteredTaxonomies = taxonomiesProp => {
//
}

// on l.80
filteredTaxonomies(taxonomies).map
```

xprazak2 commented:
Almost there. The dropdown list still expands and contracts as I type in the field:

![compressed](https://user-images.githubusercontent.com/2664090/74713642-4439d900-5229-11ea-9224-2f4f7a4b17ca.png)

![extended](https://user-images.githubusercontent.com/2664090/74713650-47cd6000-5229-11ea-885c-813384ec3034.png)

Maybe adding margin/padding to the filed could improve the visuals as well...

stejskalleos commented:
Updated, can't reproduce the issue with changing width of input (tested on Chrome, Opera & Firefox).
Added some margins and width limit to look better.

xprazak2 commented:
Thanks!
